pep-9999.rst

```diff
@@ -0,0 +1,506 @@
+
+An empty requirement table ``{}`` places no restriction on the requirement.
+
+optional-dependencies
+---------------------
+
+Format: table
+
+The keys inside this table are the names of an extra's required distribution.
+The values can have one of the following types:
+
+- table: a requirement table.
+
+- array: an array of requirement tables.
+
+These requirement tables have
+`the same specification as above <#requirement-spec>`_, with the addition of
+the following required key:
+
+- ``for-extra`` (string): the name of the :pep:`508` extra that this
+  requirement is required for.
+
+
+Reference implementation
+========================
+
+Tools will need to convert this format to :pep:`508` requirement strings. Below
+is an example implementation of that conversion (assuming validation is already
+performed):
+
+.. code-block::
+
+    def convert_requirement_to_pep508(name, requirement):
+        if isinstance(requirement, str):
+            requirement = {"version": requirement}
+        pep508 = name
+        if "extras" in requirement:
+            pep508 += " [" + ", ".join(requirement["extras"]) + "]"
+        if "version" in requirement:
+            pep508 += " " + requirement["version"]
+        if "url" in requirement:
+            pep508 += " @ " + requirement["url"]
+        for vcs in ("git", "hg", "bzr", "svn"):
+            if vcs in requirement:
+                pep508 += " @ " + vcs + "+" requirement[vcs]
+                if "revision" in requirement:
+                    pep508 += "@" + revision
+        extra = None
+        if "for-extra" in requirement:
+            extra = requirement["for-extra"]
+        if "markers" in requirement:
+            markers = requirement["markers"]
+            if extra:
+                markers = "extra = '" + extra + "' and (" + markers + ")"
+            pep508 += "; " + markers
+        return pep508, extra
+
+
+    def convert_requirements_to_pep508(dependencies):
+        pep508s = []
+        extras = []
+        for name, req in dependencies.items():
+            if isinstance(req, list):
+                for sub_req in req:
+                    pep508, extra = convert_requirement_to_pep508(name, sub_req)
+                    pep508s.append(pep508)
+                    if extra:
+                        extras.append(extra)
+            else:
+                pep508, extra = convert_requirement_to_pep508(name, sub_req)
+                pep508s.append(pep508)
+                if extra:
+                    extras.append(extra)
+        return pep508s, extras
+
+
+    def convert_project_requirements_to_pep508(project):
+        reqs, _ = convert_requirements_to_pep508(project.get("dependencies", {}))
+        optional_reqs, extras = convert_requirements_to_pep508(
+            project.get("optional-dependencies", {})
+        )
+        reqs += optional_reqs
+        return reqs, extras
+
+
+Examples
+========
+
+Full example:
```

Comment:
Is this a real-life example or a made-up one? If the former can you link to the original? If the latter then it might be worth mentioning that this is artificial for demonstration purposes.